### ISSUE TITLE
mutagen doesn't install mutagen agents

### ISSUE THREAD

**Issue by alexvorobiev** [NixOS/nixpkgs#81219]
**Describe the bug**
Trying to run mutagen results in an error:
```
$ mutagen sync resume bedc2b69-b8ef-42c3-98bc-993059c84d7b
Started Mutagen daemon in background (terminate with "mutagen daemon stop")     
Extracting agent...                                                             
Error: resume failed: unable to resume session: unable to connect to beta: unable to connect to endpoint: unable to dial agent endpoint: unable to install agent: unable to get agent for platform: unable to open agent bundle: open /nix/store/n5nx3n2wkxbpm7nfdl7w7rbllafxamqy-mutagen-0.11.0/bin/mutagen-agents.tar.gz: no such file or directory
```
The mutagen-agents.tar.gz which is supposed to have binaries for different platforms is missing.

**To Reproduce**
1. Install mutagen
2. Activate a mutagen session

**Metadata**

- system: `"x86_64-linux"`
 - host os: `Linux 5.4.20, NixOS, 20.09pre213994.e2b4abe3c8f (Nightingale)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.3`
 - channels(alex): `"home-manager"`
 - channels(root): `"nixos-20.09pre213994.e2b4abe3c8f, home-manager"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:mutagen
# a list of nixos modules affected by the problem
module:
```

**Comment by havoc-io** [comment]
Hey, author of Mutagen here, super excited to see a NixOS package for it!

I just thought I'd mention that the `mutagen-agents.tar.gz` bundle currently needs to be placed in the same directory as the `mutagen` executable itself.  This is something I've been meaning to improve for a while now (by allowing it to be put in a `libexec` directory), but I haven't had a chance to implement it yet, mostly because it's a little tricky to get right with package managers like Homebrew that symlink binaries into a different target location.

I'm not familiar with NixOS's package management, but I'm sure that it would prefer to avoid having this tarball in the `bin` directory.  So I'm happy to push ahead on trying to support the `libexec`-based layout if that's helpful here (or perhaps something more general if NixOS prefers a different layout).

**Comment by veprbl** [comment]
cc @marsam

**Comment by havoc-io** [comment]
A quick update:  Mutagen v0.11.2 and later support looking for the `mutagen-agents.tar.gz` bundle in a `libexec` directory (assuming a Filesystem Hierarchy Standard layout), so a layout of the form:

```
bin/
    mutagen
libexec/
    mutagen-agents.tar.gz
```

should now work (in addition to the option of installing the bundle directly alongside the executable).

This also works in a [Homebrew](https://brew.sh/)-style setup where the `mutagen` binary is symlinked into a larger `bin` directory.

**Comment by alexvorobiev** [comment]
@havoc-io Thanks! Is the `mutagen-agents.tar.gz` available for direct download?

**Comment by havoc-io** [comment]
@alexvorobiev The `mutagen-agents.tar.gz` bundle ships as part of the official releases (e.g. in the [`mutagen_linux_amd64_v0.11.5.tar.gz`](https://github.com/mutagen-io/mutagen/releases/download/v0.11.5/mutagen_linux_amd64_v0.11.5.tar.gz) bundle available [here](https://github.com/mutagen-io/mutagen/releases/tag/v0.11.5)).  It needs to be kept in sync with the version of the `mutagen` binary being installed, but it should work just fine with a `mutagen` binary built locally, so long as the versions are the same (to the patch level).

The `mutagen-agents.tar.gz` bundle can also be built locally, but doing so requires running a custom script inside the Mutagen source tree that performs the necessary cross-compilation (this bundle is a collection of cross-compiled binaries that get injected to remote systems to facilitate file synchronization and network forwarding).  I'm not sure the best way of handling that in the context of the NixOS packaging system.  Perhaps there are some custom hooks available?  At the moment, this script doesn't have a flag for building *just* this bundle (it will only be built as part of a full set of release bundles for all platforms), though I suppose such a flag could be added.

One of the downsides of doing the cross-compilation locally on Linux (or any non-macOS system) is that the C FFI bindings needed for efficient filesystem watching on macOS won't be available, so file synchronization between NixOS systems and macOS systems will suffer.  That might not be a huge loss, though.

This is why the Homebrew packages just use the official pre-built release bundles as their "source code."  It's a bit non-idiomatic, but it saves on local build times and allows for cross-compiled binaries with C FFI support.  Of course, I totally understand if that doesn't work with the NixOS package philosophy, in which case just building the bundle locally is probably the best bet.

**Comment by havoc-io** [comment]
As a reference point, here's what the Mutagen Homebrew formula looks like: https://github.com/mutagen-io/homebrew-mutagen/blob/master/Formula/mutagen.rb

**Comment by alexvorobiev** [comment]
I agree, setting up the cross-compilation would not be trivial so the nix derivation must download the prebuilt agents binaries from somewhere while still being able to build the agent on the platform it is running on. I see three possibilities: 

1. (No changes on the mutagen side). Nix derivation downloads the full mutagen release and extracts `mutagen-agents.tar.gz`.
2. Add the version number to the `mutagen-agents.tar.gz` and add it to the downloads. The nix derivation would just download it.
3. Provide the individual downloads for the agents similarly to `agent_linux_amd64_v0.12.0-alpha1.tar.gz` for all the platforms and have the nix derivation create the `mutagen-agents.tar.gz` itself. That way it would be very easy to even parametrize it to only download the agents the user wants and the nix derivation would supply the binary for the platform it is running on.

**Comment by havoc-io** [comment]
All three options are reasonable ideas, though I would prefer to avoid (3) since it would significantly increase the number of assets attached to each GitHub release (which is probably already too high).  The individual agent binaries for linux/amd64 that you mentioned are sort of a special case for some experimental container integration features, so I'd prefer to avoid expanding them to every platform.  I also don't know if many users want to go through the trouble of filtering the remote platform support that they want, though if that's something users are interested in then I'm certainly willing to try to support that.<sup>1</sup>

Option (2) is possible without too much effort, though I don't think it adds much value over option (1).

Additionally, if option (1) is already going to be used to download the release bundle, it might as well just pull the `mutagen` binary out of the release bundle as well and save the complexity of a local build.

In any case, the current package should have the build of `cmd/mutagen-agent` removed since it won't really be of any use.

Let me know if option (1) will work and if there's anything I can do to help.

---

<sup>1</sup> I think a better long-term solution is to provide two download variants: a "full" agent bundle that includes all platforms and a "slim" agent bundle that just includes commonly used platforms.

**Comment by bfortz** [comment]
Given the usual problems with pre-compiled binaries, reading the discussion above I thought we might have another problem : nixos needing its own agent in mutagen-agents.tar.gz. But to my surprise, the amd64 binaries work on nixos without the need to patch the ELF interpreter or anything else, it seems that everything is self-contained with no external dependencies. So why not make a binary installation directly?

**Comment by simensen** [comment]
@xenoscopic @orivej I can open a brand new issue about this, but wanted to get a quick check from you about a problem I'm having with this.

I'm getting the following when extracting an agent:

```
Extracting agent...
Error: unable to connect to beta: unable to connect to endpoint: unable to dial agent endpoint: unable to install agent: unable to get agent for platform: unable to locate agent bundle (search paths: [/Users/beausimensen/.nix-profile/bin /Users/beausimensen/.nix-profile/bin/nix/store/kir3243y2wqsxz762pmj0kvd5b7hk2c7-mutagen-0.16.0/libexec])
```

The libexec path appears to be appended to the `.nix-profile/bin` directory for some reason. I don't know if this is an issue w/ nixpkgs or mutagen but would be happy to help track it down and (maybe) fix it.

```
beausimensen@Beaus-MacBook-Pro whrc-portal % ls -l /Users/beausimensen/.nix-profile/bin/mutagen
lrwxr-xr-x 1 root wheel 70 Dec 31  1969 /Users/beausimensen/.nix-profile/bin/mutagen -> /nix/store/kir3243y2wqsxz762pmj0kvd5b7hk2c7-mutagen-0.16.0/bin/mutagen
beausimensen@Beaus-MacBook-Pro whrc-portal % ls -l /Users/beausimensen/.nix-profile/bin/nix/store/kir3243y2wqsxz762pmj0kvd5b7hk2c7-mutagen-0.16.0/libexec
ls: cannot access '/Users/beausimensen/.nix-profile/bin/nix/store/kir3243y2wqsxz762pmj0kvd5b7hk2c7-mutagen-0.16.0/libexec': No such file or directory
beausimensen@Beaus-MacBook-Pro whrc-portal % ls -l /nix/store/kir3243y2wqsxz762pmj0kvd5b7hk2c7-mutagen-0.16.0/libexec
total 0
lrwxr-xr-x 1 root wheel 87 Dec 31  1969 mutagen-agents.tar.gz -> /nix/store/nvyj8id029ighslr27jwb3plrrsm94gk-mutagen-agents-0.16.0/mutagen-agents.tar.gz
beausimensen@Beaus-MacBook-Pro whrc-portal % ls -l /nix/store/nvyj8id029ighslr27jwb3plrrsm94gk-mutagen-agents-0.16.0/mutagen-agents.tar.gz
-r--r--r-- 1 root nixbld 77M Dec 31  1969 /nix/store/nvyj8id029ighslr27jwb3plrrsm94gk-mutagen-agents-0.16.0/mutagen-agents.tar.gz
```

So, it looks like the file is there, but the search path is treating the absolute path for `/nix/store/...` as if it was relative to the `bin` path.

I suspect it has to do with the liberal usage of symlinks in nix tripping something up. I'm just too new to golang and nix to have any idea how to fix this.

**Comment by xenoscopic** [comment]
Hey @simensen, I believe this is fixed in Mutagen [v0.16.2](https://github.com/mutagen-io/mutagen/releases/tag/v0.16.2) (thanks to @dj95, who reported it in mutagen-io/mutagen#400 and fixed it in mutagen-io/mutagen#401).  It looks like the corresponding NixOS package needs to be updated — we're currently at Mutagen v0.16.4.  If you want to open a NixOS/nixpkgs PR for that, that's probably the way to go.  Feel free to @ mention me in that PR, otherwise I'll see if I can find some time to open it during this week.

**Comment by simensen** [comment]
@xenoscopic Thanks for the information. :) I spent some time yesterday trying to get my local system running on Mutagen 0.16.4. I got it working in terms of correctly locating the agents, but I noted in the PR that it isn't working for docker containers because Mutagen running as root does not seem to be able to find Docker installed via nix.

I'm happy to move any additional discussion to #215142.
